File: app/src/main/java/com/imangazalievm/bubbble/data/shots/ShotsRepository.java
package com.imangazalievm.bubbble.data.shots;

import com.imangazalievm.bubbble.Constants;
import com.imangazalievm.bubbble.data.global.network.Dribbble;
import com.imangazalievm.bubbble.data.global.network.DribbbleApiService;
import com.imangazalievm.bubbble.domain.global.models.Shot;
import com.imangazalievm.bubbble.domain.global.models.ShotsRequestParams;
import com.imangazalievm.bubbble.domain.global.models.ShotsSearchRequestParams;
import com.imangazalievm.bubbble.domain.global.models.UserShotsRequestParams;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.Single;

public class ShotsRepository {

    private DribbbleApiService dribbbleApiService;
    private DribbbleSearchDataSource dribbbleSearchDataSource;

    @Inject
    public ShotsRepository(
            DribbbleApiService dribbbleApiService,
            DribbbleSearchDataSource dribbbleSearchDataSource
    ) {
        this.dribbbleApiService = dribbbleApiService;
        this.dribbbleSearchDataSource = dribbbleSearchDataSource;
    }

    public Single<List<Shot>> getShots(ShotsRequestParams requestParams) {
        String sort = requestParams.getSort().equals(Constants.SHOTS_SORT_POPULAR) ?
                Dribbble.Shots.type_popular : Dribbble.Shots.type_recent;

        return dribbbleApiService.getShots(sort, requestParams.getPage(), requestParams.getPageSize());
    }

    public Single<Shot> getShot(long shotId) {
        return dribbbleApiService.getShot(shotId);
    }

    public Single<List<Shot>> getUserShots(UserShotsRequestParams requestParams) {
        return dribbbleApiService.getUserShots(
                requestParams.getUserId(),
                requestParams.getPage(),
                requestParams.getPageSize()
        );
    }

    public Single<List<Shot>> search(ShotsSearchRequestParams requestParams) {
        return dribbbleSearchDataSource.search(
                requestParams.getSearchQuery(),
                requestParams.getSort(),
                requestParams.getPage(),
                requestParams.getPageSize()
        );
    }

}

File: app/src/main/java/com/imangazalievm/bubbble/data/shots/ShotsRepository.kt
package com.imangazalievm.bubbble.data.shots

import com.imangazalievm.bubbble.Constants
import com.imangazalievm.bubbble.data.global.network.Dribbble
import com.imangazalievm.bubbble.data.global.network.DribbbleApi
import com.imangazalievm.bubbble.data.global.parsing.PageParserManager
import com.imangazalievm.bubbble.domain.global.models.Shot
import com.imangazalievm.bubbble.domain.global.models.ShotsRequestParams
import com.imangazalievm.bubbble.domain.global.models.ShotsSearchParams
import com.imangazalievm.bubbble.domain.global.models.UserShotsRequestParams
import javax.inject.Inject

class ShotsRepository @Inject constructor(
    private val dribbbleApi: DribbbleApi,
    private val pageParserManager: PageParserManager,
    private val searchPageParser: SearchPageParser
) {

    suspend fun getShots(requestParams: ShotsRequestParams): List<Shot> {
        val sort =
            if (requestParams.sort == Constants.SHOTS_SORT_POPULAR)
                Dribbble.Shots.Type.POPULAR
            else Dribbble.Shots.Type.PERSONAL
        return dribbbleApi.getShots(sort.code, requestParams.page, requestParams.pageSize)
    }

    suspend fun getShot(shotId: Long): Shot {
        return dribbbleApi.getShot(shotId)
    }

    suspend fun getUserShots(requestParams: UserShotsRequestParams): List<Shot> {
        return dribbbleApi.getUserShots(
            requestParams.userId,
            requestParams.page,
            requestParams.pageSize
        )
    }

    suspend fun search(params: ShotsSearchParams): List<Shot> {
        return pageParserManager.parse(searchPageParser, params)
    }

}